File: Week 3/python/14.binary_to_decimal.py
"""
WEEK 3 - PYTHON LOOPS & NUMBER THEORY
Topic: Binary to Decimal Conversion
File: 14.binary_to_decimal.py

PROBLEM:
Read an integer that LOOKS LIKE a binary number (e.g. 1100) and output its
decimal value (e.g. 12).

CONCEPT:
Iterate digits right to left, accumulate digit * 2^position into a sum.
Equivalent to evaluating the polynomial in base 2.

KEY POINTS:
 - `n % 10` -> rightmost "binary" digit (0 or 1)
 - `n //= 10` -> drop the rightmost digit
 - `pow *= 2` -> next position
 - Built-in: `int("1100", 2)` -> 12  (parse string in given base)

SYNTAX:
 dec = 0; pow = 1
 while n:
     dec += (n % 10) * pow
     pow *= 2
     n //= 10

DRY RUN (1100):
 n=1100 last=0 pow=1 -> dec=0;  n=110, pow=2
 n=110  last=0       -> dec=0;  n=11,  pow=4
 n=11   last=1       -> dec=4;  n=1,   pow=8
 n=1    last=1       -> dec=12; n=0    STOP
 -> 12

COMPLEXITY: O(d) where d = number of binary digits.
"""

import sys


def binary_to_decimal(n: int) -> int:
    dec = 0
    pow_ = 1
    while n > 0:
        bit = n % 10
        if bit not in (0, 1):
            raise ValueError(f"non-binary digit: {bit}")
        dec += bit * pow_
        pow_ *= 2
        n //= 10
    return dec


def main() -> None:
    if sys.stdin.isatty():
        for n in (1100, 1, 0, 11111111, 10101010):
            print(f"{n:>10} (binary) -> {binary_to_decimal(n)} (decimal)   "
                  f"[built-in: {int(str(n), 2)}]")
        return
    n = int(sys.stdin.read().split()[0])
    print(binary_to_decimal(n))


if __name__ == "__main__":
    main()


# NOTES:
# - Built-in `int(s, 2)` parses a binary STRING; cleaner than digit math.
# - For 32+ bit binary numbers, you might exceed int range in Java; Python is fine.
# - Conversely, `bin(12)` -> '0b1100' (string with 0b prefix).

File: Week 3/cpp/14.binary_to_decimal.cpp
/*
 * WEEK 3 - C++ LOOPS & NUMBER THEORY
 * Topic: Binary to Decimal Conversion
 * File: 14.binary_to_decimal.cpp
 *
 * PROBLEM:
 *  Read a base-10 integer that LOOKS LIKE binary (e.g. 1100), output its
 *  decimal value (12).
 *
 * KEY POINTS:
 *  - Peel last digit with `% 10`, multiply by power of 2, accumulate.
 *  - Built-in alternative: `std::stoi("1100", nullptr, 2)` parses base 2.
 *
 * DRY RUN:
 *  1100 -> 0*1 + 0*2 + 1*4 + 1*8 = 12
 */

#include <iostream>
#include <string>

long long bin_to_dec(long long n) {
    long long dec = 0, p = 1;
    while (n > 0) {
        int bit = n % 10;
        dec += bit * p;
        p *= 2;
        n /= 10;
    }
    return dec;
}

int main() {
    long long n;
    if (!(std::cin >> n)) {
        for (long long v : {1100LL, 1LL, 0LL, 11111111LL}) {
            std::cout << v << " (binary) -> " << bin_to_dec(v)
                      << "    [stoi: " << std::stoll(std::to_string(v), nullptr, 2) << "]\n";
        }
        return 0;
    }
    std::cout << bin_to_dec(n) << '\n';
    return 0;
}

/*
 * NOTES:
 *  - For Java's Integer.parseInt(s, 2), C++ uses std::stoi/std::stoll with a radix.
 *  - Always validate that each "digit" is 0 or 1 in production code.
 */
